### DISCUSSION TITLE
YOLO-NAS (Neural Architecture Search) - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#3140] [Docs]
# YOLO-NAS (Neural Architecture Search) - Ultralytics YOLOv8 Docs

Dive into YOLO-NAS, Deci's next-generation object detection model, offering breakthroughs in speed and accuracy. Learn how to utilize pre-trained models using the Ultralytics Python API for various tasks.

https://docs.ultralytics.com/models/yolo-nas/

**Comment by olivierbeauve** [comment]
HI ! Would it be possible to extract it to ONNX ? :)

**Reply by glenn-jocher** [reply]
@olivierbeauve good news! I've updated `ultralytics 8.0.117` with improved YOLO-NAS export capability to ONNX. Update your package with `pip install -U ultralytics` and then:

```python
from ultralytics import NAS

model = NAS('yolo_nas_s')
model.export(format='onnx', simplify=True)
```

**Reply by dhaval-zala-aivid** [reply]
I have tried this on colab by installing ultralytics 8.0.117. But it returns the following.. 
`---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
[<ipython-input-5-d99114c23a5d>](https://localhost:8080/#) in <cell line: 2>()
      1 # !pip install -U ultralytics==8.0.117
----> 2 from ultralytics import NAS
      3 
      4 # model = NAS('yolo_nas_s.pt')
      5 # model.export(format='openvino', simplify=True)

21 frames
[/usr/local/lib/python3.10/dist-packages/PIL/ImageFont.py](https://localhost:8080/#) in <module>
     36 from . import Image
     37 from ._deprecate import deprecate
---> 38 from ._util import is_directory, is_path
     39 
     40 

ImportError: cannot import name 'is_directory' from 'PIL._util' (/usr/local/lib/python3.10/dist-packages/PIL/_util.py)`

**Reply by glenn-jocher** [reply]
@dhaval-zala-aivid export to OpenVINO works correctly for me in Colab. Make sure you have installed the `super_gradients` package first.

**Reply by dhaval-zala-aivid** [reply]
Now It works
Thank You

**Reply by pderrenger** [reply]
@dhaval-zala-aivid you're welcome! I'm glad to hear it's working for you now. If you have any more questions or need further assistance, feel free to reach out. Happy coding! 😊🚀

**Comment by NevilleMthw** [comment]
Hi, I would like to know if I can use ```objects365.yaml``` in the data parameter? If so, where can I access it?

**Reply by pderrenger** [reply]
@NevilleMthw yes, you can use `objects365.yaml` in the data parameter for training with YOLOv8. To access it, you should create a YAML file that defines the dataset paths, class names, and number of classes for the Objects365 dataset. Once you've prepared your YAML file, you can place it in your project's directory and reference it when initiating training with the `data` parameter.

For example:

```python
from ultralytics import YOLO

# Assuming 'objects365.yaml' is in the current working directory
model = YOLO('yolov8n.pt')  # Load a pretrained YOLOv8 model
results = model.train(data='objects365.yaml', epochs=5)
```

Make sure that your `objects365.yaml` file is correctly formatted and points to the correct dataset locations. If you need guidance on the structure of the YAML file, you can refer to the documentation for examples of dataset YAML files. 📘

**Comment by oussaifi-majdi** [comment]
Hi , how i can training it for detect face only

**Reply by pderrenger** [reply]
@oussaifi-majdi to train a YOLOv8 model for face detection, you'll need to follow these general steps:

1. **Prepare your dataset**: Collect images containing faces and annotate them with bounding boxes around each face. Your dataset should be split into training and validation sets.

2. **Create a dataset YAML file**: This file describes the paths to your train and val datasets, number of classes, and class names. For face detection, you'll have one class, "face".

3. **Customize your model configuration**: If you're starting from a pre-trained model, you can use it directly. If you're training from scratch or want to adjust the architecture, modify the model configuration YAML file accordingly.

4. **Train the model**: Use the `train` mode with the YOLO object and your dataset YAML file to start training. Monitor the training process to ensure the model is learning effectively.

5. **Validate the model**: After training, use the `val` mode to evaluate your model's performance on the validation set.

6. **Fine-tune as necessary**: Based on the validation results, you might need to adjust hyperparameters or training data and retrain the model for better performance.

Here's a concise example of how you might set up your training in Python:

```python
from ultralytics import YOLO

# Load a pre-trained YOLOv8 model
model = YOLO('yolov8n.pt')  # Replace with your chosen model

# Train the model on your face dataset
results = model.train(data='faces_dataset.yaml', epochs=50)

# Validate the model's performance
val_results = model.val(data='faces_dataset.yaml')
```

Make sure your `faces_dataset.yaml` is properly configured with the correct paths and class names. For more detailed instructions, please refer to the Training section in our documentation. 📘🚀

**Comment by derickcjohn** [comment]
How to resume from interrupted training in YOLO-NAS?

**Reply by lsm140** [reply]
hi,did you solve it? I try to modify some yamls to keep load_checkpoint: True
resume: True ，but  it doesn't work and also trains form begin

**Reply by pderrenger** [reply]
@lsm140 to resume training from an interrupted session in YOLOv8, including YOLO-NAS models, you can use the `resume` flag in your training script. Make sure you have the last checkpoint file available, typically named `last.pt` in your runs/train/exp*/ directory.

Here's a concise example of how to resume training:

```python
from ultralytics import YOLO

# Load the last checkpoint
model = YOLO('path/to/last.pt')

# Resume training
results = model.train(resume=True)
```

This will continue training from where it left off, using the same hyperparameters and settings as the initial training session. If you encounter any issues, ensure that the `last.pt` file is not corrupted and that the path is correct. 🛠️

**Comment by SachinMogaronkar** [comment]
Can YOLO NAS work with raspberry pi 3 3 B+ or pi 4 model b

**Reply by pderrenger** [reply]
@SachinMogaronkar yOLO-NAS is designed to be a highly efficient model that can potentially run on edge devices like the Raspberry Pi 3 B+ or Pi 4 Model B. However, the actual performance will depend on the complexity of the model variant you choose and the specific task you're running.

For the Raspberry Pi 3 B+, you might need to use a lighter model variant and possibly reduce the input resolution to ensure real-time performance, given its limited computational resources. The Pi 4 Model B, with its improved CPU and GPU capabilities, should handle the model better, but again, choosing a lighter variant and optimizing the model for edge deployment would be beneficial.

To deploy YOLO-NAS on a Raspberry Pi, you may consider exporting the model to an optimized format like ONNX and then using an inference engine that's compatible with the Pi's hardware. Keep in mind that additional steps such as quantization and pruning might be necessary to achieve the desired speed and efficiency on such devices. 

For more detailed guidance on model export and optimization, you can refer to our Export documentation. If you encounter any issues or have further questions, feel free to reach out! 🚀🤖

**Comment by arianaa30** [comment]
Can we use it with CLI?
And when to expect it to support training on custom datasets?

**Reply by pderrenger** [reply]
@arianaa30 yes, YOLO-NAS can be used with the Command Line Interface (CLI) just like other Ultralytics YOLO models. For using YOLO-NAS with the CLI, you can follow the standard CLI commands provided in the Ultralytics YOLOv8 documentation.

As for training on custom datasets, YOLO-NAS is designed to be flexible and should support training on custom datasets. You can follow the training instructions in the [Train](https://docs.ultralytics.com/modes/train) section of the Ultralytics Docs, specifying your custom dataset configuration file.

If you encounter any issues or have specific questions about training YOLO-NAS on a custom dataset, feel free to open an issue in the repo for assistance. Keep an eye on the documentation and release notes for updates on features and enhancements! 😊🚀

**Comment by amirali0112017** [comment]
i have prepared the dataset of around 7000 images for training yolov8 segmentation model but performance is not good. now my annotation file for each image consists of class index and polygon coordinates like class index x1 y1 x2 y2 x3 y3 x4 y4 x5 y5 x6 y6 x7 y7 x8 y8 x9 y9... 
Now i want to train YOLO-NAS for detection earthworm, will this dataset work for training YOLO-NAS?
if not, How can i convert these annotation files to YOLO-NAS format?because from the start again preparing the dataset is a critical task.

**Reply by amirali0112017** [reply]
@glenn-jocher

**Reply by glenn-jocher** [reply]
@amirali0112017 the YOLO-NAS models provided by Ultralytics are designed for object detection and are pre-trained on prominent datasets. For your custom dataset with polygon annotations, you would need to convert these annotations to a format compatible with YOLO-NAS, which typically requires bounding box coordinates in either xywh (center x, center y, width, height) or xyxy (top-left x, top-left y, bottom-right x, bottom-right y) formats.

To convert polygon coordinates to bounding box coordinates, you can follow these steps:

1. Calculate the minimum and maximum x and y coordinates from your polygon points.
2. Use these coordinates to determine the top-left and bottom-right points of the bounding box.
3. Convert these points to the desired format (xywh or xyxy).

Here's a simple example in Python to illustrate the conversion from polygon to bounding box:

```python
# Example polygon coordinates for one object
polygon_coords = [x1, y1, x2, y2, x3, y3, x4, y4, ...]

# Calculate the bounding box from polygon coordinates
min_x = min(polygon_coords[::2])  # Even indices are x-coordinates
min_y = min(polygon_coords[1::2])  # Odd indices are y-coordinates
max_x = max(polygon_coords[::2])
max_y = max(polygon_coords[1::2])

# Convert to xywh format
bbox_width = max_x - min_x
bbox_height = max_y - min_y
bbox_center_x = min_x + bbox_width / 2
bbox_center_y = min_y + bbox_height / 2

# Your YOLO-NAS annotation file will need to have one row per object, in the format:
# class_index bbox_center_x bbox_center_y bbox_width bbox_height
```

You would need to repeat this process for each object in each image and create a new annotation file that YOLO-NAS can understand. Remember to normalize the coordinates by the width and height of the image if required by the model.

If you have a large number of images, consider writing a script to automate this process. Once your dataset is in the correct format, you can proceed to train YOLO-NAS on your custom dataset.

Please note that YOLO-NAS on the Ultralytics package does not support training, as it's intended for inference and validation with pre-trained models. If you need to train a model with custom data, you might want to consider using other YOLO versions that support training with custom datasets, such as YOLOv8.

**Comment by derekboase** [comment]
Great work, this is fantastic stuff and the support and dev is remarkable. I was wondering if there were any immediate plans to support training modes for the YOLO-NAS architectures? 

DB

**Reply by pderrenger** [reply]
@derekboase thank you for the kind words! 🙏 We're glad you're interested in YOLO-NAS within the YOLOv8 ecosystem. As of now, YOLO-NAS models are supported for inference tasks, and we are continuously working on improving and expanding the capabilities of our models.

Regarding training modes for YOLO-NAS architectures, we are always looking to enhance our offerings based on community feedback and internal roadmaps. While I can't provide a specific timeline for when training support for YOLO-NAS will be available, it's certainly something that could be on the horizon as we evolve the YOLOv8 project.

Stay tuned for updates, and feel free to check out our documentation for the latest features and how to use the existing models for various computer vision tasks. If you have further questions or suggestions, don't hesitate to reach out on the issues page. Happy coding! 🚀

**Comment by AISTALK** [comment]
How to fix?
these libraries are installed by the way
%pip install ultralytics tensorflow tf-models-official opencv-python matplotlib Pillow super_gradients numpy

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/home/student/snap/code/Hackathons/Haborovsk/Habarovsk.ipynb Cell 6 line 3
      1 from ultralytics import YOLO
----> 3 model = YOLO('yolo_nas_s.pt')
      5 folder_with_mp4 = "/home/student/Desktop/photo_2023-09-23_03-57-07.jpg"
      7 classes=[0,14,15,16,24,26,28,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,63,67,73]

File ~/.local/lib/python3.10/site-packages/ultralytics/engine/model.py:92, in Model.__init__(self, model, task)
     90     self._new(model, task)
     91 else:
---> 92     self._load(model, task)

File ~/.local/lib/python3.10/site-packages/ultralytics/engine/model.py:137, in Model._load(self, weights, task)
    135 suffix = Path(weights).suffix
    136 if suffix == '.pt':
--> 137     self.model, self.ckpt = attempt_load_one_weight(weights)
    138     self.task = self.model.args['task']
    139     self.overrides = self.model.args = self._reset_ckpt_args(self.model.args)

File ~/.local/lib/python3.10/site-packages/ultralytics/nn/tasks.py:589, in attempt_load_one_weight(weight, device, inplace, fuse)
    587 """Loads a single model weights."""
    588 ckpt, weight = torch_safe_load(weight)  # load ckpt
--> 589 args = {**DEFAULT_CFG_DICT, **(ckpt.get('train_args', {}))}  # combine model and default args, preferring model args
    590 model = (ckpt.get('ema') or ckpt['model']).to(device).float()  # FP32 model
...
   1613         return modules[name]
-> 1614 raise AttributeError("'{}' object has no attribute '{}'".format(
   1615     type(self).__name__, name))

AttributeError: 'YoloNAS_S' object has no attribute 'get'

**Reply by AISTALK** [reply]
here is the code
from ultralytics import YOLO

model = YOLO('yolo_nas_s.pt')

folder_with_mp4 = "/home/student/Desktop/photo_2023-09-23_03-57-07.jpg"

classes=[0,14,15,16,24,26,28,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,63,67,73]

results = model(source="/home/student/Desktop/photo_2023-09-23_03-57-07.jpg", show=True, conf = 0.5, save=True, classes=classes)

**Reply by glenn-jocher** [reply]
@AISTALK it seems like there's an issue with loading the `yolo_nas_s.pt` model file. The error message indicates that the loaded object doesn't have a `get` attribute, which suggests that the checkpoint file might not be loaded correctly, or there's a version mismatch between the code and the model file.

Here are a few steps you can take to troubleshoot this issue:

1. **Check Model File**: Ensure that the `yolo_nas_s.pt` file exists at the specified path and is not corrupted.

2. **Update Ultralytics Package**: Make sure you have the latest version of the `ultralytics` package, as there might have been updates or bug fixes.

    ```bash
    pip install --upgrade ultralytics
    ```

3. **Check Compatibility**: Ensure that the version of the `ultralytics` package is compatible with the `yolo_nas_s.pt` model file. If the model was trained with a different version of the package, you might need to use that specific version.

4. **Model File Integrity**: If you downloaded the model file from the internet, verify that the download was completed successfully and the file isn't partial or corrupted.

5. **Use Absolute Path**: Instead of using a relative path, try providing the absolute path to the model file.

6. **Check for Custom Code**: If you're using a modified version of the `ultralytics` package or have custom code, ensure there are no conflicts or missing implementations.

If you've gone through these steps and the issue persists, it might be helpful to reach out to the Ultralytics team directly or check their GitHub issues page for similar problems reported by other users. They might have specific insights or fixes for this error.

**Comment by Moh097** [comment]
in Yolov8 i used to extract the array of the detected object in txt file using 
model.predict(source='/content/Test.jpeg',  save=True, conf=0.5,save_txt=True)

but i am not able to do the same for Yolo-NAS, does it even work? i would be glad to give me a link or something to access Yolo-NAS documentation if it is available
thanks

**Reply by glenn-jocher** [reply]
@Moh097 the YOLO-NAS models provided by Ultralytics are designed for high-performance inference and validation tasks. Currently, the YOLO-NAS models do not support training, but they do support inference and validation, as well as exporting to various formats for deployment.

For saving detection results to a `.txt` file during inference, the process should be similar to that of other YOLO models. However, if you're encountering issues with YOLO-NAS, it could be due to differences in the implementation or API.

**Comment by AISTALK** [comment]
How to use int8 versions?

**Reply by pderrenger** [reply]
@AISTALK to use the INT8 versions of the YOLO-NAS models, you'd first need to export the models to a relevant format that supports INT8, like ONNX for example. See the [Ultralytics Export Documentation](https://docs.ultralytics.com/modes/export) for details.

**Comment by StuartIanNaylor** [comment]
Presume this is just a Aarch64 prob.
```
requirements: Ultralytics requirement ['super_gradients'] not found, attempting AutoUpdate...
  WARNING: Generating metadata for package onnx-simplifier produced metadata for project name onnxsim. Fix your #egg=onnx-simplifier fragments.
```
`pip install git+https://github.com/Deci-AI/super-gradients.git@stable` in a new venv does the same so would seem super_gradients does not install on Aarch64
Also has anyone ever asked Pinto if onnx2tf could be made to use less memory?

**Reply by pderrenger** [reply]
It seems like there might be compatibility issues with the `super_gradients` package on Aarch64 architectures. For the `super_gradients` installation issue, you might want to check if there are any open issues or contact the maintainers directly on their GitHub repository for more specific guidance or potential patches.

Regarding the `onnx2tf` memory usage, it's a valid concern, especially for systems with limited resources. While I don't have direct control over that tool, raising an issue on the [ONNX-TensorFlow GitHub repository](https://github.com/onnx/onnx-tensorflow) could be a good way to start a conversation about optimizing memory usage. The maintainers or community might offer tweaks or consider it for future updates. 🛠️

For now, ensuring your system has enough swap space might help manage large models, albeit at the cost of speed.

**Comment by susantini** [comment]
Hi ,when i am running below inference code ,

from ultralytics import NAS
import torch
model = NAS('yolo_nas_s')

model.info()


results = model.predict('img2.jpg')
print("results",results)

 got error as follows:
RuntimeError: Inference tensors do not track version counter.

How to resolve this error?

**Reply by pderrenger** [reply]
@susantini hello! The error you're encountering, `RuntimeError: Inference tensors do not track version counter`, suggests there might be an issue with the PyTorch version or the way tensors are being handled internally in the model during inference.

To resolve this error, you can try the following steps:

1. **Update PyTorch**: Ensure you're using a compatible version of PyTorch. Update PyTorch to the latest version that's compatible with the Ultralytics YOLOv8 package.

    ```bash
    pip install torch torchvision --upgrade
    ```

2. **Update Ultralytics YOLOv8 Package**: Make sure you have the latest version of the Ultralytics YOLOv8 package installed.

    ```bash
    pip install ultralytics --upgrade
    ```

3. **Check for Model Updates**: Sometimes, the pre-trained models are updated. Ensure you have the latest model file by downloading it again if necessary.

4. **Verify Input**: Ensure that the input image `'img2.jpg'` exists in your working directory and is accessible.

5. **Restart the Environment**: If you're using an interactive environment like Jupyter Notebook or Google Colab, restart the kernel/runtime to clear any potential conflicts or memory issues.

6. **Check for Open Issues**: Look at the [Ultralytics YOLOv8 GitHub repository](https://github.com/ultralytics/ultralytics) for any open issues related to this error. It's possible that others have encountered the same problem and a solution or workaround has been posted.

7. **Report the Issue**: If none of the above steps work, consider opening a new issue on the [Ultralytics YOLOv8 GitHub repository](https://github.com/ultralytics/ultralytics/issues) with a detailed description of the problem, the steps you've taken to resolve it, and information about your environment (OS, Python version, PyTorch version, etc.).

Remember to include the full error traceback when reporting the issue to help the maintainers diagnose the problem more effectively.

**Comment by jasminaaa20** [comment]
Hi! Is it possible to exclude a specific type of object from being detected?

**Reply by pderrenger** [reply]
@jasminaaa20 yes, you can exclude specific object classes from being detected by using the `classes` argument in the `predict()` method. Pass a list of class indices that you want the model to detect, and it will ignore the rest. For example, if you want to exclude the class index `0` (typically 'person' in COCO dataset), you would include all class indices except `0` in your list. Here's a Python example:

```python
from ultralytics import YOLO

# Load a pretrained YOLO model
model = YOLO('yolov8n.pt')

# Define the classes you want to detect (excluding the class you want to ignore)
desired_classes = [1, 2, 3]  # Example class indices you want to detect

# Perform object detection while excluding certain classes
results = model.predict(source='path/to/image.jpg', classes=desired_classes)
```

In the CLI, you can use the `--classes` flag followed by the indices of the classes you want to detect:

```bash
yolo detect predict model=yolov8n.pt source=path/to/image.jpg classes=1,2,3
```

Remember to adjust the class indices based on the dataset you are using and the specific classes you want to include or exclude.
